token/client/src/token.rs

```diff
@@ -2261,6 +2263,33 @@ where
             )
         };
 
+        let (transfer_amount_auditor_ciphertext_lo, transfer_amount_auditor_ciphertext_hi) =
+            if let Some(proof_data) = ciphertext_validity_proof_data {
+                let transfer_amount_auditor_ciphertext_lo = proof_data
+                    .context_data()
+                    .grouped_ciphertext_lo
+                    .try_extract_ciphertext(2)
+                    .map_err(|_| TokenError::ProofGeneration)?;
+                let transfer_amount_auditor_ciphertext_hi = proof_data
+                    .context_data()
+                    .grouped_ciphertext_hi
+                    .try_extract_ciphertext(2)
+                    .map_err(|_| TokenError::ProofGeneration)?;
+                (
+                    transfer_amount_auditor_ciphertext_lo,
+                    transfer_amount_auditor_ciphertext_hi,
+                )
+            } else {
+                // the validity proof data is always generated unless
+                // `transfer_amount_auditor_ciphertext_lo` and
+                // `transfer_amount_auditor_ciphertext_hi` are `Some`, so it is
+                // safe to unwrap
+                (
+                    *transfer_amount_auditor_ciphertext_lo.unwrap(),
+                    *transfer_amount_auditor_ciphertext_hi.unwrap(),
+                )
```

Comment:
I don't think this is guaranteed by the function (unless we add a new type as I suggested), so we should return an error if either is None